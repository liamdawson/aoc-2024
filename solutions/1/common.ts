const extractorRegex = /^\s*(\d+)\s+(\d+)\s*$/;
export function getSortedColumns(contentLines: string[]): [number[], number[]] {
  const lists: [number[], number[]] = [[], []];
  const addValues = (values: string[]) => {
    lists[0].push(Number(values[0]));
    lists[1].push(Number(values[1]));
  };

  for (const line of contentLines) {
    const extracted = extractorRegex.exec(line);
    if (!extracted) {
      continue;
    }
    const [_line, ...values] = extracted;

    addValues(values);
  }

  for (const list of lists) {
    list.sort((a, b) => a - b);
  }

  return lists;
}
