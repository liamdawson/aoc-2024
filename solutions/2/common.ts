export function arrayWithoutItemAt<T>(array: T[], index: number): T[] {
  return array.slice(0, index).concat(array.slice(index + 1));
}

export function reportIsSafe(
  report: number[],
  options: { dampener?: boolean } = {},
): boolean | undefined {
  // not covered by specification
  if (report.length < 2) {
    return undefined;
  }

  const ascending = report[0] < report[1];
  const [min, max] = ascending ? [1, 3] : [-3, -1];

  for (let i = 0; i < report.length - 1; i++) {
    const a = report[i];
    const b = report[i + 1];
    const diff = b - a;

    if (diff < min || diff > max) {
      if (options?.dampener) {
        // try removing all of the levels up to the last one we're comparing - removing later levels won't help
        for (let j = 0; j < Math.min(report.length, i + 2); j++) {
          if (reportIsSafe(arrayWithoutItemAt(report, j))) {
            return true;
          }
        }
      }

      return false;
    }
  }

  return true;
}

export function parseReports(inputLines: string[]) {
  return inputLines.flatMap((line) => {
    const report = line
      .split(" ")
      .filter((v) => v.trim())
      .map((value) => Number(value))
      .filter((n) => !Number.isNaN(n));

    if (report.length > 0) {
      return [report];
    }

    return [];
  });
}
