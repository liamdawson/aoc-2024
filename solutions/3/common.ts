interface Instructions {
  mul: [number, number];
}
export type InstructionOperation = keyof Instructions;

export type Instruction<
  Operation extends InstructionOperation = InstructionOperation,
> = {
  readonly [Op in InstructionOperation]: {
    op: Op;
    args: Instructions[Op];
  };
}[Operation];

export type ParsedInstruction<
  Operation extends InstructionOperation = InstructionOperation,
> = Instruction<Operation> & { raw: string };

const knownInstructions = ["mul"] as const;
const maybeInstruction = new RegExp(
  `(${knownInstructions.join("|")})(?=([(][^)]*[)]))`,
  "g",
);
const remainderRegexFor: Record<string, RegExp> = {
  mul: /^\((\d+),(\d+)\)$/,
};

export function parseInstructions(input: string): ParsedInstruction[] {
  const mulMatches = input
    .matchAll(maybeInstruction)
    .flatMap(([raw, start, remainder]) => {
      const op = start as InstructionOperation;

      const remainderRegex = remainderRegexFor[op];
      const remainderResult = remainderRegex?.exec(remainder);

      if (!remainderResult) {
        return [];
      }

      if (op === "mul") {
        const [_, arg1, arg2] = remainderResult;
        const args: [number, number] = [Number(arg1), Number(arg2)];

        if (Number.isNaN(args[0]) || Number.isNaN(args[1])) {
          return [];
        }

        return [
          {
            op,
            args,
            raw,
          },
        ] as const;
      }

      return [];
    });

  return [...mulMatches];
}

export function evaluateInstruction(instruction: Instruction) {
  switch (instruction.op) {
    case "mul":
      return instruction.args[0] * instruction.args[1];
  }

  // type error on unknown instruction
  instruction.op satisfies never;
  throw new Error("Unknown operation");
}
