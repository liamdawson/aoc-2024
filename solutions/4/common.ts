// north = "up" (earlier rows), west = "left" (earlier columns)
const searchDirections = {
  northwest: { colStep: -1, rowStep: -1 },
  north: { colStep: 0, rowStep: -1 },
  northeast: { colStep: 1, rowStep: -1 },
  west: { colStep: -1, rowStep: 0 },
  east: { colStep: 1, rowStep: 0 },
  southwest: { colStep: -1, rowStep: 1 },
  south: { colStep: 0, rowStep: 1 },
  southeast: { colStep: 1, rowStep: 1 },
} as const;
const allSearchDirections = Object.keys(
  searchDirections,
) as readonly SearchDirection[];
export type SearchDirection = keyof typeof searchDirections;

interface SearchGrid {
  letters: string[][];
  width: number;
  height: number;
  stringFrom(
    start: Point,
    takeNext: number,
    direction: SearchDirection,
  ): string | null;
}

export type Point = [row: number, column: number];

export function gridify(lines: string[]): SearchGrid {
  const letters: string[][] = [];
  const width = lines[0].length;

  for (const line of lines) {
    const row = line.trim().split("");
    if (row.length !== width) {
      throw new Error("Not all lines are of equal width!");
    }

    letters.push(row);
  }

  return {
    letters,
    width,
    height: letters.length,
    stringFrom(start: Point, takeNext: number, direction: SearchDirection) {
      const [startRow, startCol] = start;
      const { rowStep, colStep } = searchDirections[direction];

      const rowEnd = startRow + rowStep * takeNext;
      const colEnd = startCol + colStep * takeNext;

      // check whether the request is in range
      if (
        rowEnd < 0 ||
        rowEnd >= letters.length ||
        colEnd < 0 ||
        colEnd >= width
      ) {
        return null;
      }

      const resultLetters = [];
      for (let i = 0; i <= takeNext; i++) {
        resultLetters.push(
          letters[startRow + rowStep * i][startCol + colStep * i],
        );
      }

      return resultLetters.join("");
    },
  };
}

export function getMatches(needle: string, haystack: SearchGrid) {
  const needleLetters = needle.split("");
  const start = needleLetters[0];
  const remainderLength = needleLetters.length - 1;
  let matchesFound = 0;

  for (let rowIndex = 0; rowIndex < haystack.height; rowIndex++) {
    const row = haystack.letters[rowIndex];

    for (let colIndex = 0; colIndex < haystack.width; colIndex++) {
      const letter = row[colIndex];

      if (letter === start) {
        const matchesHere = allSearchDirections
          .map((direction) =>
            haystack.stringFrom(
              [rowIndex, colIndex],
              remainderLength,
              direction,
            ),
          )
          .filter((result) => result === needle).length;

        matchesFound += matchesHere;
      }
    }
  }

  return matchesFound;
}
