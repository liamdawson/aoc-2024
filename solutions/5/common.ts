export interface PageOrderRequirement {
  pageNumber: number;
  mustPrecede: number;
}

export type PageUpdateList = number[];

const pageOrderRequirementRegex = /^\s*(\d+)[|](\d+)\s*$/g;
export function parsePageOrderRequirement(
  line: string,
): PageOrderRequirement | null {
  const parseResult = pageOrderRequirementRegex.exec(line);
  if (!parseResult) {
    return null;
  }

  const [raw, rawPageNumber, rawMustPrecede] = parseResult;
  const [pageNumber, mustPrecede] = [
    Number(rawPageNumber),
    Number(rawMustPrecede),
  ];

  if (Number.isNaN(pageNumber) || Number.isNaN(mustPrecede)) {
    throw new Error(`Failed to parse "${raw}" as a page order requirement`);
  }

  return { pageNumber, mustPrecede };
}

export function parsePageUpdateList(line: string): PageUpdateList {
  const pageNumbers = line.split(",").map((pageNumber) => Number(pageNumber));

  if (pageNumbers.some((number) => Number.isNaN(number))) {
    throw new Error(`Failed to parse page update list ${line}`);
  }

  return pageNumbers;
}

export function parseInput(lines: string[]) {
  const pageOrderRequirements = [];
  const pageUpdateLists = [];
  let requirementsSection = true;

  while (true) {
    if (lines.length === 0) {
      break;
    }

    const line = lines.shift();

    if (line?.trim() === "") {
      requirementsSection = false;
      continue;
    }

    if (line === undefined) {
      continue;
    }

    if (requirementsSection) {
      pageOrderRequirements.push(parsePageOrderRequirement(line));
    } else {
      pageUpdateLists.push(parsePageUpdateList(line));
    }
  }

  return {
    pageOrderRequirements: pageOrderRequirements.filter(
      Boolean,
    ) as PageOrderRequirement[],
    pageUpdateLists,
  };
}

export const makeVerifier = (pageOrderRequirements: PageOrderRequirement[]) => {
  const mustPrecedes = Map.groupBy(
    pageOrderRequirements,
    (req) => req.pageNumber,
  );

  return (pageUpdateList: PageUpdateList) => {
    const seen = new Set();

    for (const pageNumber of pageUpdateList) {
      const disallowed =
        mustPrecedes.get(pageNumber)?.map(({ mustPrecede }) => mustPrecede) ??
        [];

      if (
        disallowed.some((restrictedPageNumber) =>
          seen.has(restrictedPageNumber),
        )
      ) {
        return {
          valid: false,
          pageUpdateList,
        } as const;
      }

      seen.add(pageNumber);
    }

    return {
      valid: true,
      pageUpdateList,
      middleNumber: pageUpdateList[Math.floor(pageUpdateList.length / 2)],
    } as const;
  };
};
