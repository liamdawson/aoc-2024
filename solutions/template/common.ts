export function todo() {
	return true;
}
