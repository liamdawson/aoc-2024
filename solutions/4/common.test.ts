import { describe, expect, test } from "vitest";
import { getMatches, gridify } from "./common.ts";

const basicGrid = `XMASS
MASXA
ASXMM
SXMAX`.split("\n");

describe(gridify.name, () => {
  test("basic grid", () => {
    expect(gridify(basicGrid)).toMatchSnapshot();
  });
});

describe(getMatches.name, () => {
  test("basic grid", () => {
    expect(getMatches("XMAS", gridify(basicGrid))).toEqual(3);
  });
});
