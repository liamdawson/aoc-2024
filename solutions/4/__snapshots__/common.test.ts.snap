// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`gridify > basic grid 1`] = `
{
  "height": 4,
  "letters": [
    [
      "X",
      "M",
      "A",
      "S",
      "S",
    ],
    [
      "M",
      "A",
      "S",
      "X",
      "A",
    ],
    [
      "A",
      "S",
      "X",
      "M",
      "M",
    ],
    [
      "S",
      "X",
      "M",
      "A",
      "X",
    ],
  ],
  "stringFrom": [Function],
  "width": 5,
}
`;
